File: SimpleTunnelSSHForLinux/tests/SimpleTunnelSSH.Core.Tests/PortConflictDetectorTests.cpp
#include "TestRegistry.h"

#include <stdexcept>

#include "Models/TunnelGroup.h"
#include "Models/TunnelPort.h"
#include "Services/PortConflictDetector.h"

namespace SimpleTunnelSSH::Core::Tests
{

namespace
{

void expect(bool condition, const char* message)
{
    if (!condition)
    {
        throw std::runtime_error(message);
    }
}

} // namespace

int runPortConflictDetectorTests()
{
    using namespace SimpleTunnelSSH::Core::Models;
    using namespace SimpleTunnelSSH::Core::Services;

    {
        TunnelGroup development;
        development.name = QStringLiteral("Development");
        development.ports = { TunnelPort { QUuid::createUuid(), {}, true, 3000, QStringLiteral("localhost"), 3000 } };

        TunnelGroup production;
        production.name = QStringLiteral("Production");
        production.ports = { TunnelPort { QUuid::createUuid(), {}, true, 3000, QStringLiteral("localhost"), 3000 } };

        const auto conflicts = PortConflictDetector::detect({ development, production });
        expect(conflicts.size() == 1, "Expected exactly one port conflict.");
        expect(conflicts.first().localPort == 3000, "Expected conflicting local port 3000.");
    }

    {
        TunnelGroup development;
        development.name = QStringLiteral("Development");
        development.ports = { TunnelPort { QUuid::createUuid(), {}, true, 3000, QStringLiteral("localhost"), 3000 } };

        TunnelGroup database;
        database.name = QStringLiteral("Database");
        database.ports = { TunnelPort { QUuid::createUuid(), {}, true, 5432, QStringLiteral("localhost"), 5432 } };

        const auto conflicts = PortConflictDetector::detect({ development, database });
        expect(conflicts.isEmpty(), "Expected no conflicts when local ports are unique.");
    }

    {
        TunnelGroup development;
        development.name = QStringLiteral("Development");
        development.ports = { TunnelPort { QUuid::createUuid(), {}, false, 3000, QStringLiteral("localhost"), 3000 } };

        TunnelGroup production;
        production.name = QStringLiteral("Production");
        production.ports = { TunnelPort { QUuid::createUuid(), {}, true, 3000, QStringLiteral("localhost"), 3000 } };

        const auto conflicts = PortConflictDetector::detect({ development, production });
        expect(conflicts.isEmpty(), "Expected disabled ports to be ignored.");
    }

    return 0;
}

} // namespace SimpleTunnelSSH::Core::Tests

File: SimpleTunnelSSHForWindows/tests/SimpleTunnelSSH.Core.Tests/PortConflictDetectorTests.cs
using SimpleTunnelSSH.Core.Models;
using SimpleTunnelSSH.Core.Services;

namespace SimpleTunnelSSH.Core.Tests;

public sealed class PortConflictDetectorTests
{
    [Fact]
    public void Detect_ReturnsConflict_WhenTwoGroupsUseTheSameLocalPort()
    {
        var development = new TunnelGroup
        {
            Name = "Development",
            Ports = [new TunnelPort { LocalPort = 3000, RemoteHost = "localhost", RemotePort = 3000 }]
        };

        var production = new TunnelGroup
        {
            Name = "Production",
            Ports = [new TunnelPort { LocalPort = 3000, RemoteHost = "localhost", RemotePort = 3000 }]
        };

        var conflicts = PortConflictDetector.Detect([development, production]);

        Assert.Single(conflicts);
        Assert.Equal(3000, conflicts[0].LocalPort);
    }

    [Fact]
    public void Detect_ReturnsEmpty_WhenPortsAreUnique()
    {
        var development = new TunnelGroup
        {
            Name = "Development",
            Ports = [new TunnelPort { LocalPort = 3000, RemoteHost = "localhost", RemotePort = 3000 }]
        };

        var database = new TunnelGroup
        {
            Name = "Database",
            Ports = [new TunnelPort { LocalPort = 5432, RemoteHost = "localhost", RemotePort = 5432 }]
        };

        var conflicts = PortConflictDetector.Detect([development, database]);

        Assert.Empty(conflicts);
    }

    [Fact]
    public void Detect_IgnoresDisabledPorts()
    {
        var development = new TunnelGroup
        {
            Name = "Development",
            IsEnabled = true,
            Ports = [new TunnelPort { LocalPort = 3000, RemoteHost = "localhost", RemotePort = 3000, IsEnabled = false }]
        };

        var production = new TunnelGroup
        {
            Name = "Production",
            IsEnabled = true,
            Ports = [new TunnelPort { LocalPort = 3000, RemoteHost = "localhost", RemotePort = 3000, IsEnabled = true }]
        };

        var conflicts = PortConflictDetector.Detect([development, production]);

        Assert.Empty(conflicts);
    }
}